pkg/kube/krt/processor.go

```diff
@@ -0,0 +1,251 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/buffer"
+
+	"istio.io/istio/pkg/slices"
+)
+
+// handlerSet tracks a set of handlers. Handlers can be added at any time.
+type handlerSet[O any] struct {
+	mu               sync.RWMutex
+	handlers         []*processorListener[O]
+	listenersStarted bool
+	wg               wait.Group
+}
+
+func (o *handlerSet[O]) Run(stopCh <-chan struct{}) {
+	func() {
+		o.mu.Lock()
+		defer o.mu.Unlock()
+		for _, listener := range o.handlers {
+			o.wg.Start(listener.run)
+			o.wg.Start(listener.pop)
+		}
+		o.listenersStarted = true
+	}()
+	<-stopCh
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, listener := range o.handlers {
+		close(listener.addCh) // Tell .pop() to stop. .pop() will tell .run() to stop
+	}
+
+	// Wipe out list of listeners since they are now closed
+	// (processorListener cannot be re-used)
+	o.handlers = nil
+
+	// Reset to false since no listeners are running
+	o.listenersStarted = false
+
+	o.wg.Wait() // Wait for all .pop() and .run() to stop
+}
+
+func (o *handlerSet[O]) Insert(f func(o []Event[O], initialSync bool), parentSynced Syncer, initialEvents []Event[O]) Syncer {
+	o.mu.Lock()
+	l := newProcessListener(f, parentSynced)
+
+	o.handlers = append(o.handlers, l)
+	if o.listenersStarted {
+		o.wg.Start(l.run)
+		o.wg.Start(l.pop)
+	}
+	o.mu.Unlock()
+	log.Debugf("send initial events (%d)", len(initialEvents))
+	l.send(initialEvents, true)
+	return l.Synced()
+}
+
+func (o *handlerSet[O]) Distribute(events []Event[O], initialSync bool) {
+	o.mu.RLock()
+	defer o.mu.RUnlock()
+	for _, listener := range o.handlers {
+		listener.send(slices.Clone(events), initialSync)
+	}
+}
+
+// Synced returns a Syncer that will wait for all registered handlers (at the time of the call!) are synced.
+func (o *handlerSet[O]) Synced() Syncer {
+	o.mu.RLock()
+	syncer := multiSyncer{syncers: make([]Syncer, 0, len(o.handlers))}
+	for _, listener := range o.handlers {
+		syncer.syncers = append(syncer.syncers, listener.Synced())
+	}
+	o.mu.RUnlock()
+	return syncer
+}
+
+// processorListener is a fork of the upstream Kubernetes one. It has minimal tweaks to fit into
+// our use case better, but the most important one being that processorListener is private in Kubernetes.
+//
+// processorListener relays notifications from a sharedProcessor to
+// one ResourceEventHandler --- using two goroutines, two unbuffered
+// channels, and an unbounded ring buffer.  The `add(notification)`
+// function sends the given notification to `addCh`.  One goroutine
+// runs `pop()`, which pumps notifications from `addCh` to `nextCh`
+// using storage in the ring buffer while `nextCh` is not keeping up.
+// Another goroutine runs `run()`, which receives notifications from
+// `nextCh` and synchronously invokes the appropriate handler method.
+//
+// processorListener also keeps track of the adjusted requested resync
+// period of the listener.
+type processorListener[O any] struct {
+	nextCh chan any
```

Comment:
why `chan any`? is there any case where it would not be an `eventSet[O]` that I've just missed?

Reply:
It was for some silly reason, plausibly not needed. I have already blocked it from my memory, will rethink it